src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,854 @@
+   ring_iso::GenRingIso
+   mat_iso::GenMatIso
+   AbstractAlgebra.@declare_other
+
+   MatrixGroup(m::Int, F::Ring) = new{elem_type(F), dense_matrix_type(elem_type(F))}(m,F)
+
+end
+
+# build a MatrixGroup given the underlying GAP object
+function MatrixGroup(m::Int, F::S, G_gap::GapObj) where S<:Ring
+   r = MatrixGroup(m,F)
+   r.X = G_gap
+   return r
+end
+
+# build a MatrixGroup given a list of generators
+function MatrixGroup(m::Int, F::S, L::Vector) where S<:Ring
+   r = MatrixGroup(m,F)
+   r.gens = L
+   for l in r.gens l.parent = r end
+   return r
+end
+
+
+# NOTE: at least one of the fields :elm and :X is always defined, but not necessarly both of them.
+"""
+    MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: GAPGroupElem{MatrixGroup}
+
+Elements of a group of type `MatrixGroup{RE<:RingElem, T<:MatElem{RE}}`
+"""
+mutable struct MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: AbstractMatrixGroupElem
+   parent::MatrixGroup{RE, T}
+   elm::T                         # Oscar matrix
+   X::GapObj                     # GAP matrix. If x isa MatrixGroupElem, then x.X = x.parent.mat_iso(x.elm)
+
+   #MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x::MatElem{RE}) where RE <: RingElem = new{RE, MatElem{RE}}(G,x)
+   MatrixGroupElem(G::MatrixGroup, x::MatElem) = new{elem_type(G.ring),dense_matrix_type(elem_type(G.ring))}(G,x)
+
+   #function MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x_gap::GapObj) where RE <: RingElem
+   function MatrixGroupElem(G::MatrixGroup, x_gap::GapObj)
+      z = new{elem_type(G.ring),dense_matrix_type(elem_type(G.ring))}()
+      z.parent = G
+      z.X = x_gap
+      return z
+   end
+
+end
+
+# build a MatrixGroupElem given its underlying GAP object
+# WARNING: this does not check whether the element actually lies in the group G
+function MatrixGroupElem(G::MatrixGroup, x::T, x_gap::GapObj)  where T <: MatElem
+   z = MatrixGroupElem(G,x)
+   z.X = x_gap
+   return z
+end
+
+
+ring_elem_type(::Type{MatrixGroup{S,T}}) where {S,T} = S
+mat_elem_type(::Type{MatrixGroup{S,T}}) where {S,T} = T
+_gap_filter(::Type{<:MatrixGroup}) = GAP.Globals.IsMatrixGroup
+
+########################################################################
+#
+# Basic
+#
+########################################################################
+
+function Base.show(io::IO, x::MatrixGroup)
+   AbstractAlgebra.@show_name(io, x)
+   AbstractAlgebra.@show_special(io,x)
+
+   if isdefined(x, :descr)
+      if x.descr==:GU || x.descr==:SU
+         print(io, string(x.descr), "(",x.deg,",",characteristic(x.ring)^(div(degree(x.ring),2)),")")
+      else
+         print(io, string(x.descr), "(",x.deg,",",order(x.ring),")")
+      end
+   else
+      print(io, "Matrix group of degree ", x.deg, " over ")
+      show(IOContext(io, :compact => true), x.ring)
+   end
+end
+
+function Base.show(io::IO, x::MatrixGroupElem)          #TODO: is this correct?
+   if isdefined(x, :elm)
+      show(io, "text/plain", x.elm)
+#      print(io, x.elm)
```

Comment:
To quote the documentation for `Base.show` on two arguments:
> The representation used by show generally includes Julia-specific formatting and type information, and should be parseable Julia code when possible.
> ...
> To customize human-readable text output for objects of type T, define show(io::IO, ::MIME"text/plain", ::T) instead.

So perhaps we should *not* do this, but instead call just plain `show(io, x.elm)`, and then add another method `Base.show(::IO, ::MIME"text/plain", : x::MatrixGroupElem)` which invokes `show(io, "text/plain", x.elm)`???

Perhaps @rfourquet has some insight into this?

Reply:
the `show` machinery is known to be difficult to understand in Julia... I would agree with @fingolfin here for not calling the 3-args `show` method within the 2-args one. And I don't know of a (simple) way of avoiding writing two different `show` methods (2- and 3-args) which just forward the presence or absence of a `MIME` parameter.

Reply:
The printing/show machinery is indeed difficult. At the moment we have the following for matrices:
```julia
julia> A = matrix(ZZ, 2, 2, [1, 2, 3, 4])
[1  2]
[3  4]

julia> [A, A]
2-element Array{AbstractAlgebra.Generic.MatSpaceElem{BigInt},1}:
 [1 2; 3 4]
 [1 2; 3 4]

julia> print(A)
[1 2; 3 4]
```
I guess we want to the same behavior? The "compact" behavior is coming from overloading
```julia
function Base.show(io::IO, a::T)
```
The ordinary REPL printing comes from overloading
```julia
function Base.show(io::IO, ::MIME"text/plain", a::T)
```